jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java

```diff
@@ -0,0 +1,1204 @@
+    private final Map<String, PathsApp> trackedApps = new HashMap<>();
+    private final Map<String, Attributes> environmentAttributesMap = new HashMap<>();
+
+    private Comparator<DeployAction> actionComparator = new DeployActionComparator();
+    private Path environmentsDir;
+    private int scanInterval = 10;
+    private Scanner scanner;
+    private boolean useRealPaths;
+    private boolean deferInitialScan = false;
+    private String defaultEnvironmentName;
+
+    /**
+     * <p>
+     * Convenience constructor to have DeploymentScanner participate in DeploymentManager startup behaviors,
+     * allowing failures in DeploymentScanner.start() to trigger a failure to start DeploymentManager.
+     * </p>
+     *
+     * @param server the server reference to use for any XML based deployments.
+     * @param deploymentManager the DeploymentManager instance that is being used by this DeploymentScanner.
+     */
+    public DeploymentScanner(
+        @Name("server") Server server,
+        @Name("deploymentManager") DeploymentManager deploymentManager)
+    {
+        this(server, deploymentManager, null);
+        deploymentManager.addBean(this);
+    }
+
+    /**
+     * <p>
+     * Construct a raw DeploymentScanner that will (periodically) scan specific directories for paths that can be
+     * used to construct webapps that will be submitted to the DeploymentManager for eventual deployment to
+     * it's configured destination.
+     * </p>
+     *
+     * @param server the server reference to use for any XML based deployments.
+     * @param contextHandlerDeployer the ContextHandlerDeployer to use for deploying the created ContextHandlers.
+     */
+    public DeploymentScanner(
+        @Name("server") Server server,
+        @Name("contextHandlerDeployer") ContextHandlerDeployer contextHandlerDeployer)
+    {
+        this(server, contextHandlerDeployer, null);
+    }
+
+    /**
+     * <p>
+     * Construct a raw DeploymentScanner that will (periodically) scan specific directories for paths that can be
+     * used to construct webapps that will be submitted to the DeploymentManager for eventual deployment to
+     * it's configured destination.
+     * </p>
+     *
+     * @param server the server reference to use for any XML based deployments.
+     * @param contextHandlerDeployer the ContextHandlerDeployer to use for deploying the created ContextHandlers.
+     * @param filter A custom FilenameFilter to control what files the {@link Scanner} monitors for changes.
+     */
+    public DeploymentScanner(
+        @Name("server") Server server,
+        @Name("contextHandlerDeployer") ContextHandlerDeployer contextHandlerDeployer,
+        @Name("filenameFilter") FilenameFilter filter)
+    {
+        this.server = server;
+        this.contextHandlerDeployer = Objects.requireNonNull(contextHandlerDeployer);
+        this.filenameFilter = Objects.requireNonNullElse(filter, new MonitoredPathFilter(monitoredDirs));
+    }
```

Comment:
This is a little fragile, but if we are doing the beans like this then this is needed so all constructors work the same:

```suggestion
    {
        this.server = server;
        this.contextHandlerDeployer = Objects.requireNonNull(contextHandlerDeployer);
        this.filenameFilter = Objects.requireNonNullElse(filter, new MonitoredPathFilter(monitoredDirs));
        if (contextHandlerDeployer instanceof Container container)
            container.addBean(this);

    }
```